transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactoryTest.java

```diff
@@ -0,0 +1,272 @@
+                will(returnValue(webSocketExtensionSpi));
+            }
+        });
+        ActiveExtensions activeExtensions =
+                wsExtFactory.negotiateWebSocketExtensions(address, session, "Sec-Websocket-Extensions",
+                        clientRequestedExtensions);
+        Assert.assertEquals("mock", activeExtensions.asList().get(0).getExtensionToken());
+        Assert.assertEquals(1, activeExtensions.asList().size());
+        context.assertIsSatisfied();
+    }
+
+    @Test
+    public void testNegotiateWebSocketExtensionsWithParameter() throws ProtocolException {
+        List<String> clientRequestedExtensions = new ArrayList<>();
+        clientRequestedExtensions.add("not-there");
+        clientRequestedExtensions.add("mock; foo=2");
+        clientRequestedExtensions.add("not-there2; bar=2");
+        context.checking(new Expectations() {
+            {
+                oneOf(mockNegotiate).negotiate(with(new ExtensionHeaderTokenMatcher("mock").withParameter("foo=2")),
+                        with(address));
+                will(returnValue(webSocketExtensionSpi));
+            }
+        });
+        ActiveExtensions activeExtensions =
+                wsExtFactory.negotiateWebSocketExtensions(address, session, "Sec-Websocket-Extensions",
+                        clientRequestedExtensions);
+        Assert.assertEquals("mock", activeExtensions.asList().get(0).getExtensionToken());
+        Assert.assertEquals("foo=2", activeExtensions.asList().get(0).getParameters().get(0).toString());
+        Assert.assertEquals(1, activeExtensions.asList().size());
+        context.assertIsSatisfied();
+    }
+
+    @Test
+    public void testNegotiateWebSocketExtensionsNotThere() throws ProtocolException {
+        List<String> clientRequestedExtensions = new ArrayList<>();
+        clientRequestedExtensions.add("notthere");
+        ActiveExtensions activeExtensions =
+                wsExtFactory.negotiateWebSocketExtensions(address, session, "Sec-Websocket-Extensions",
+                        clientRequestedExtensions);
+        Assert.assertTrue(activeExtensions.asList().isEmpty());
+    }
+
+    @Test
+    public void testNegotiatedWebSocketExtensionRejectedParameter() throws ProtocolException {
+        List<String> clientRequestedExtensions = new ArrayList<>();
+        clientRequestedExtensions.add("mock");
+        context.checking(new Expectations() {
+            {
+                oneOf(mockNegotiate).negotiate(with(new ExtensionHeaderTokenMatcher("mock")), with(address));
+                will(returnValue(null));
+            }
+        });
+        ActiveExtensions activeExtensions =
+                wsExtFactory.negotiateWebSocketExtensions(address, session, "Sec-Websocket-Extensions",
+                        clientRequestedExtensions);
+        Assert.assertEquals(0, activeExtensions.asList().size());
+        context.assertIsSatisfied();
+    }
+
```

Comment:
Need to tests for 
- case where WebSocketExtensionFactorySpi.negotiate throws ProtocolException (for negotiate and negotiateWebSocketExtensions)
- more than one supported WebSocket extension (order in response header should be same as order in request header for extensions that much support in ones)